### ISSUE TITLE
Allow accessing inner classes on instances for annotations

### ISSUE THREAD

**Issue by freundTech** [python/mypy#13826]
**Feature**

It would be very useful if mypy would allow accessing inner classes (and type aliases) on instances when used in annotations. Currently this is not allowed:

```python
class A:
    class B: ...
    
a: A = A()
b: a.B = a.B()
```
Accessing the inner class on the outer class is allowed:
```python
class A:
    class B: ...
    
a: A = A()
b: A.B = a.B()
```

**Pitch**

I'm currently trying to write a stub generator for cffi. cffi is a Python-C-binding-generator.
A cffi generated python module contains two variables: `lib` of type `_cffi_backend.Lib`, which exposes the bound C functions and `ffi` of type `_cffi_backend.FFI`, which exposes various utility function.

`_cffi_backend.FFI` also contains multiple inner classes that users of cffi will often want to use, such as `CData`.

If this feature is implemented in mypy, users would be able to type code using cffi as such:
```python
from cffi_generated_module import ffi, lib

def foo(a: ffi.CData) -> None:
    pass
```
Currently mypy complains about `Name "ffi.CData" is not defined`, requiring the user to manually import _cffi_backend.FFI instead.
Note that during untyped usage the user will never have to directly interact with `_cffi_backend` and probably doesn't even know that it exists.

**PEP**

As far as I can tell neither PEP-484, nor any other PEP comments on this. This is neither explicitly allowed nor forbidden.

**Possible problems**

Allowing this might lead to problems with inheritance:
```python
class A:
    class Inner: ...
    
class B(A):
    class Inner: ...
    
a: A
i: a.Inner = a.Inner()
```
It's not statically knowable is b is of type `A.Inner` or `B.Inner`, which are unrelated types.
One possible solution would be to require `B.Inner` to inherit from `A.Inner`. I haven't looked into if this breaks existing code (mypy primer or other) yet.
However I think it should be ok to just ignore this problem and always assume the inner type of the static type (in this case `A.Inner`).
The following is already allowed my mypy and also not correct:

```python
class A:
    class Inner: ...
    
class B(A):
    class Inner: ...
    
a: A = B()
i: A.Inner = a.Inner()
```

**Comment by erictraut** [comment]
Use of variables within a type annotation isn't allowed in the Python type system. For example, the statement `b: a.B` in the top example isn't permitted because `a` is a variable.

There are good reasons why variables are not allowed in type annotations. The type of variable `a` at that point in the code flow could be a union of types. It could also depend on type narrowing from assignments or conditional expressions elsewhere in the code flow.

Your other examples above are allowed because `A.B` and `A.Inner` can be statically evaluated with no code flow analysis required. They unambiguously refer to a specific type.

**Comment by thomkeh** [comment]
Do you actually need to define the inner class within the other class? Otherwise you could do this:
```python
from typing_extensions import TypeAlias

class Inner_: ...

class A:
    Inner: TypeAlias = Inner_

x: A.Inner = A.Inner()
reveal_type(x)  # N: Revealed type is "inner_class.Inner_"
```

**Comment by freundTech** [comment]
> Use of variables within a type annotation isn't allowed in the Python type system. For example, the statement `b: a.B` in the top example isn't permitted because `a` is a variable.

Ah. Right. I searched the PEPs specifically for mentions of inner classes, so I missed the bigger problem...

> Do you actually need to define the inner class within the other class? Otherwise you could do this:

The problem is that the user doesn't have access to the outer class, but just to an instance of it, so they can't reference anything - neither an inner class nor a TypeAlias - on the outer class (unless they imports an underscored module `_cffi_backend` that they usually never needs to interact with)

I ended up working around the problem my introducing a new module `cffi_types`, that just contains TypeAliases to the types from the underscored module. This has the disadvantage that the types need to be present at runtime (unless the user uses an `if TYPE_CHECKING` guard for the import and uses `from __future__ import annotations`), but is probably good enough.
